src/main/java/spoon/support/reflect/declaration/CtModuleImpl.java

```diff
@@ -0,0 +1,446 @@
+		if (moduleMember == null) {
+			return (T) this;
+		}
+
+		return this.addModuleMemberAt(this.moduleMembers.size(), moduleMember);
+	}
+
+	private CtRole computeRoleFromModuleMember(CtModuleMember moduleMember) {
+		CtRole role;
+		if (moduleMember instanceof CtModuleRequirement) {
+			role = CtRole.REQUIRED_MODULE;
+		} else if (moduleMember instanceof CtUsedService) {
+			role = CtRole.SERVICE_TYPE;
+		} else if (moduleMember instanceof CtProvidedService) {
+			role = CtRole.PROVIDED_SERVICE;
+		} else if (moduleMember instanceof CtPackageExport) {
+			CtPackageExport packageExport = (CtPackageExport) moduleMember;
+			if (packageExport.isOpenedPackage()) {
+				role = CtRole.OPENED_PACKAGE;
+			} else {
+				role = CtRole.EXPORTED_PACKAGE;
+			}
+		} else {
+			role = CtRole.MODULE_MEMBER;
+		}
+		return role;
+	}
+
+	@Override
+	public <T extends CtModule> T addModuleMemberAt(int position, CtModuleMember moduleMember) {
+		if (moduleMember == null) {
+			return (T) this;
+		}
+
+		if (this.moduleMembers == CtElementImpl.<CtModuleMember>emptyList()) {
+			this.moduleMembers = new SortedList<>(new CtLineElementComparator());
+		}
+		if (!this.moduleMembers.contains(moduleMember)) {
+			moduleMember.setParent(this);
+			CtRole role = this.computeRoleFromModuleMember(moduleMember);
+
+			getFactory().getEnvironment().getModelChangeListener().onListAdd(this, role, this.moduleMembers, position, moduleMember);
+			this.moduleMembers.add(position, moduleMember);
+		}
+
+		return (T) this;
+	}
+
+	@Override
+	public List<CtModuleMember> getModuleMembers() {
+		return Collections.unmodifiableList(this.moduleMembers);
+	}
+
+	@Override
+	public <T extends CtModule> T removeModuleMember(CtModuleMember moduleMember) {
+		if (moduleMember == null || this.moduleMembers.size() == 0) {
+			return (T) this;
+		}
+		if (this.moduleMembers.contains(moduleMember)) {
+			getFactory().getEnvironment().getModelChangeListener().onListDelete(this, this.computeRoleFromModuleMember(moduleMember), this.moduleMembers, this.moduleMembers.indexOf(moduleMember), moduleMember);
+			if (this.moduleMembers.size() == 1) {
+				this.moduleMembers = CtElementImpl.emptyList();
+			} else {
+				this.moduleMembers.remove(moduleMember);
+			}
+		}
+
+		return (T) this;
+	}
+
+	@Override
+	public <T extends CtModule> T setIsOpenModule(boolean openModule) {
+		getFactory().getEnvironment().getModelChangeListener().onObjectUpdate(this, CtRole.MODIFIER, openModule, this.openModule);
+		this.openModule = openModule;
+		return (T) this;
+	}
+
+	@Override
+	public List<CtUsedService> getUsedServices() {
+		if (this.moduleMembers.isEmpty()) {
+			return CtElementImpl.emptyList();
+		} else {
+			List<CtUsedService> usedServices = new ArrayList<>();
+			for (CtModuleMember moduleMember : this.moduleMembers) {
+				if (moduleMember instanceof CtUsedService) {
+					usedServices.add((CtUsedService) moduleMember);
+				}
+			}
+			return usedServices;
```

Comment:
I suggest that all such geters return unmodifiable collection - to make it clear to clients that it is not read/write collection.